sdk/servicebus/service-bus/src/connectionContext.ts

```diff
@@ -354,47 +401,46 @@ export namespace ConnectionContext {
+      if (!state.wasConnectionCloseCalled && state.numReceivers) {
+        logger.verbose(
+          `[${connectionContext.connection.id}] connection.close() was not called from the sdk and there were ${state.numReceivers} ` +
+            `receivers. We should reconnect.`
+        );
+
+        // Call onDetached() on receivers so that batching receivers it can gracefully close any ongoing batch operation
+        await callOnDetachedOnReceivers(
+          connectionContext,
+          connectionError || contextError,
+          "batching"
+        );
+        await callOnDetachedOnReceivers(
+          connectionContext,
+          connectionError || contextError,
+          "session"
+        );
+      }
+
       await refreshConnection(connectionContext);
       waitForConnectionRefreshResolve();
       waitForConnectionRefreshPromise = undefined;
       // The connection should always be brought back up if the sdk did not call connection.close()
       // and there was at least one receiver link on the connection before it went down.
       logger.verbose("[%s] state: %O", connectionContext.connectionId, state);
+
+      // Calling onDetached on streaming receiver
       if (!state.wasConnectionCloseCalled && state.numReceivers) {
         logger.verbose(
           `[${connectionContext.connection.id}] connection.close() was not called from the sdk and there were ${state.numReceivers} ` +
             `receivers. We should reconnect.`
         );
-        await delay(Constants.connectionReconnectDelay);
-
-        const detachCalls: Promise<void>[] = [];
-
-        // Call onDetached() on receivers so that batching receivers it can gracefully close any ongoing batch operation
-        // and streaming receivers can decide whether to reconnect or not.
-        for (const receiverName of Object.keys(connectionContext.messageReceivers)) {
-          const receiver = connectionContext.messageReceivers[receiverName];
-          if (receiver) {
-            logger.verbose(
-              "[%s] calling detached on %s receiver '%s'.",
-              connectionContext.connection.id,
-              receiver.receiverType,
-              receiver.name
-            );
-            detachCalls.push(
-              receiver.onDetached(connectionError || contextError).catch((err) => {
-                logger.logError(
-                  err,
-                  "[%s] An error occurred while calling onDetached() on the %s receiver '%s'",
-                  connectionContext.connection.id,
-                  receiver.receiverType,
-                  receiver.name
-                );
-              })
-            );
-          }
-        }
 
-        await Promise.all(detachCalls);
+        // Call onDetached() on streaming receivers so that they can recover
+        await callOnDetachedOnReceivers(
+          connectionContext,
+          connectionError || contextError,
+          "streaming"
```

Comment:
Calling onDetached on streaming receivers after the `refreshConnection` since `onDetached` would recover the streaming receiver and that would only be possible after the refresh connection.

Reply:
Walking through this mentally it seems like it should be fine. I believe the sequence we'll end up with is something like this:

'connection refreshing promise started'
onDetach called on all receivers
   batchingReceiver.onDetach
      batchingReceiver.closeLink
         batchingReceiver.openLock is now locked
   <ad infinitum>
connection refresh initiates
connection closed
new connection created
refresh connection promise resolved
  (continues) batchingReceiver.closeLink
     batchingReceiver.openLock is now **acquired**
     batchingReceiver.closeLink completes (against old connection, but that's fine
     batchingReceiver.openLink happens (against new connection, which is what you intend)

Does that match your analysis?

Reply:
One correction..

The `closeLink` ends completely and then the refresh is initiated.

Reply:
Thanks for sharing this reasoning! Any chance you can put it as a comment as well for our future selves?

Reply:
Added more comments. :)